src/winbind.c

```diff
@@ -206,15 +206,78 @@ uint32_t winbind_cli_auth(char *user, char *domain,
     return ret;
 }
 
+const char gssntlmssp_sids_urn[] = "urn:gssntlmssp:sids";
+
+static uint32_t format_sids_as_name_attribute(
+                    const struct wbcAuthUserInfo *wbc_info,
+                    struct gssntlm_name_attribute **auth_attrs)
+{
+    char *sids_buf;
+    char *sids_buf_realloced;
+    size_t offset = 0;
+    struct gssntlm_name_attribute *attrs;
+    size_t worst_sids_list_len = WBC_SID_STRING_BUFLEN * wbc_info->num_sids;
+
+    /* Allocate buffers */
+
+    /* 1 for returned attribute +1 for termiator entry */
+    attrs = calloc(2, sizeof(struct gssntlm_name_attribute));
+    if (attrs == NULL) {
+        return ENOMEM;
+    }
+
+    /* sids buffer is allocated with the worst-case size */
+    sids_buf = malloc(worst_sids_list_len);
+    if (sids_buf == NULL) {
+        free(attrs);
+        return ENOMEM;
+    }
+
+    /* Construct name attributes string */
+    for (uint32_t i = 0; i < wbc_info->num_sids; i++) {
+        offset += wbcSidToStringBuf(&wbc_info->sids[i].sid, sids_buf + offset,
+                                    worst_sids_list_len - offset);
+        if (i < wbc_info->num_sids - 1) {
+            /* Replace EOL by separator for non-last SID */
+            sids_buf[offset] = ',';
+        }
+        offset++;
+    }
+
+    /* Usually average SID has ~5 sub_authorities out of 15 possible so
+     * about 60% of worst-case string size is unused.  Having 100 SIDs
+     * in ACCESS_TOKEN, the space waste is about 11k out of 18k.
+     * Optimization: to save the space, we shrink sids_buf here */
+    sids_buf_realloced = realloc(sids_buf, offset);
```

Comment:
I'm skeptical that this will have the effect you claim.  It's up to the malloc implementation whether to give you a different region or not.  In particular, glibc's malloc uses an arena allocator at this size, and will not - which means there's no gain here.

This whole thing seems like premature optimization.  If space usage is actually a concern here, you may wish to do this the other way around and use a buffer that you realloc larger as needed.

Reply:
(See malloc.c lines 3230-3232.)

Reply:
But I have already done two-pass algorithm before (see previous commit) and @simo5  have asked to use realloc to save the space instead...?

From realloc I don't actually need a DIFFERENT region, it's ok to give the same.
I need that that the last allocated region is shrinked and the tail is reused for the next allocation (of smaller size).
The tail is 11k in average so it can contain even large blocks of 4k.

The 'use a buffer that you realloc larger as needed' has O(N^2) complexity and definitely not acceptable here (each time we can't enlarge the buffer in-place we need to move old content of the buffer to new place). Sids list can vary in size in very wide range (from a hundred of bytes to 190k) so we simply can't do it with such poor algorithm.

Do you see this behaviour in GLIBC (malloc.c)? Then probably this is an impediment and we need to return to two-pass algorithm OR allocate a second smaller buffer and copy constructed blob part bit-wise to that second buffer and then release a big (the first) one. Bitwise copy of large block can be optimized and probably better than two-pass generation of SIDs string (using wbcSidToStringBuf->sprintf).

Reply:
I am ok with this code as is (even w/o the realloc actually). If we find that this is a problem in practice we'll deal with it.

Reply:
Ok, I still want to emphasize that memory waste is ~60% in average (11k of 18k per auth) and this is too much for stress load. If you decide to remove realloc, please leave at least my comment so we can return to this place in case of issues.

Reply:
I am ok with the code as is. I believe malloc will properly reclaim and reuse the rest of the allocated memory in later calls. And if malloc doesn't I believe it is a bug in malloc.